File: fe/app/models/permission.js
import Ember from 'ember';
import DS from 'ember-data';

export default DS.Model.extend({
  user: DS.belongsTo('user'),

  active: DS.attr(),

  user_read: DS.attr(),
  user_create: DS.attr(),
  user_edit: DS.attr(),
  user_delete: DS.attr(),

  player_read: DS.attr(),
  player_create: DS.attr(),
  player_edit: DS.attr(),
  player_delete: DS.attr(),

  unit_read: DS.attr(),
  unit_create: DS.attr(),
  unit_edit: DS.attr(),
  unit_delete: DS.attr(),
  unit_apply: DS.attr(),
  unit_accept: DS.attr(),
  unit_assign: DS.attr(),

  ship_model_read: DS.attr(),
  ship_model_create: DS.attr(),
  ship_model_edit: DS.attr(),
  ship_model_delete: DS.attr(),

  ship_read: DS.attr(),
  ship_create: DS.attr(),
  ship_edit: DS.attr(),
  ship_delete: DS.attr(),

  reward_read: DS.attr(),
  reward_create: DS.attr(),
  reward_edit: DS.attr(),
  reward_delete: DS.attr(),

  //   inserted_at: DS.attr(),
  //   created_at: DS.attr(),


  settings: Ember.computed('user_read', 'ship_read', 'reward_read', function() {
    return this.get('user_read') || this.get('ship_read') || this.get('reward_read');
  }),

  ship_list_filter: Ember.computed('ship_model_read', 'ship_read', function() {
    return this.get('user_read') || this.get('ship_read') || this.get('reward_read');
  }),

  ship_tb: Ember.computed('ship_edit', 'ship_delete', function() {
    return this.get('ship_edit') || this.get('ship_delete');
  }),
});

File: be/web/models/permission.ex
defmodule OrgtoolDb.Permission do
  use OrgtoolDb.Web, :template
  alias OrgtoolDb.Repo

  schema "permissions" do

    field :active,        :boolean, default: false

    field :user_read,     :boolean, default: false
    field :user_create,   :boolean, default: false
    field :user_edit,     :boolean, default: false
    field :user_delete,   :boolean, default: false

    field :player_read,   :boolean, default: false
    field :player_create, :boolean, default: false
    field :player_edit,   :boolean, default: false
    field :player_delete, :boolean, default: false

    field :unit_read,     :boolean, default: false
    field :unit_create,   :boolean, default: false
    field :unit_edit,     :boolean, default: false
    field :unit_delete,   :boolean, default: false
    field :unit_apply,    :boolean, default: false
    field :unit_accept,   :boolean, default: false
    field :unit_assign,   :boolean, default: false

    field :ship_model_read,     :boolean, default: false
    field :ship_model_create,   :boolean, default: false
    field :ship_model_edit,     :boolean, default: false
    field :ship_model_delete,   :boolean, default: false

    field :ship_read,     :boolean, default: false
    field :ship_create,   :boolean, default: false
    field :ship_edit,     :boolean, default: false
    field :ship_delete,   :boolean, default: false

    field :reward_read,   :boolean, default: false
    field :reward_create, :boolean, default: false
    field :reward_edit,   :boolean, default: false
    field :reward_delete, :boolean, default: false

    belongs_to :user, OrgtoolDb.User

    timestamps()
  end

  def preload do
    [:user]
  end

  def all!(permission) do
    permission
    |> Repo.preload(:user)
    |> changeset(
      %{
        active: true,
        user_read: true, user_create: true, user_edit: true, user_delete: true,
        player_read: true, player_create: true, player_edit: true, player_delete: true,
        unit_read: true, unit_create: true, unit_edit: true, unit_delete: true, unit_apply: true, unit_accept: true, unit_assign: true,
        ship_model_read: true, ship_model_create: true, ship_model_edit: true, ship_model_delete: true,
        ship_read: true, ship_create: true, ship_edit: true, ship_delete: true,
        reward_read: true, reward_create: true, reward_edit: true, reward_delete: true
      })
      |> Repo.update!
  end


  @permissions [
    :active,
    :user_read, :user_create, :user_edit, :user_delete,
    :player_read, :player_create, :player_edit, :player_delete,
    :unit_read, :unit_create, :unit_edit, :unit_delete, :unit_apply, :unit_accept, :unit_assign,
    :ship_model_read, :ship_model_create, :ship_model_edit, :ship_model_delete,
    :ship_read, :ship_create, :ship_edit, :ship_delete,
    :reward_read, :reward_create, :reward_edit, :reward_delete]
  @doc """
  Builds a changeset based on the `struct` and `params`.
  """
  def changeset(struct, params \\ %{}) do
    struct
    |> cast(params, @permissions)
    |> cast_assoc(:user)
    |> validate_required(@permissions)
  end

  def changeset_include(struct, params \\ %{}) do
    struct
    |> cast(params, @permissions)
    |> validate_required(@permissions)
  end
end
